### PR TITLE
fix: add optimistic UI for deleting member on org user table

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

Fixes: CAL-4294 (partial will handle booking fixes in a new PR)

CleanShot 2024-09-16 at 10 13 14

## Mandatory Tasks (DO NOT REMOVE)

- [X] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
Load organization members
Delete a member notice they get removed from the list.

### CODE REVIEW

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx [hunk 1]
```diff
@@ -13,15 +13,39 @@ export function DeleteMemberModal({ state, dispatch }: { state: State; dispatch:
...
-        utils.viewer.teams.get.invalidate(),
-        utils.viewer.eventTypes.invalidate(),
-        utils.viewer.organizations.listMembers.invalidate(),
-      ]);
+      const previousValue = utils.viewer.organizations.listMembers.getInfiniteData({
+        limit: 10,
+        searchTerm: "",
+      });
+
+      console.log("previousValue", previousValue);
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/DeleteMemberModal.tsx:21]:
```suggestion
```

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx [hunk 2]
```diff
@@ -13,15 +13,37 @@ export function DeleteMemberModal({ state, dispatch }: { state: State; dispatch:
...
-      // We don't need to wait for invalidate to finish
-      Promise.all([
-        utils.viewer.teams.get.invalidate(),
-        utils.viewer.eventTypes.invalidate(),
-        utils.viewer.organizations.listMembers.invalidate(),
-      ]);
+      const previousValue = utils.viewer.organizations.listMembers.getInfiniteData({
+        limit: 10,
+        searchTerm: "",
+      });
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/DeleteMemberModal.tsx:19]:
```suggestion
```

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx [hunk 3]
```diff
@@ -13,15 +13,33 @@ export function DeleteMemberModal({ state, dispatch }: { state: State; dispatch:
...
+          ...oldData,
+          pages: oldData.pages.map((page) => ({
+            ...page,
+            rows: page.rows.filter((member) => member.id !== state.deleteMember.user?.id),
+          })),
+        };
+      });
+
+      // Existing invalidations
+      Promise.all([utils.viewer.teams.get.invalidate(), utils.viewer.eventTypes.invalidate()]);
```
Comment (joeauyeung) [inline packages/features/users/components/UserTable/DeleteMemberModal.tsx:29]: Do we still need to invalidate these queries as well if the previous member is removed?

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx [hunk 4]
```diff
@@ -13,15 +13,33 @@ export function DeleteMemberModal({ state, dispatch }: { state: State; dispatch:
...
-        utils.viewer.organizations.listMembers.invalidate(),
-      ]);
+      // Update the infinite data to remove the deleted user
+      utils.viewer.organizations.listMembers.setInfiniteData({ limit: 10, searchTerm: "" }, (oldData) => {
+        if (!oldData) return oldData;
+        return {
+          ...oldData,
+          pages: oldData.pages.map((page) => ({
+            ...page,
+            rows: page.rows.filter((member) => member.id !== state.deleteMember.user?.id),
```
Comment (joeauyeung) [inline packages/features/users/components/UserTable/DeleteMemberModal.tsx:23]: Stacking array methods makes me question things but I can't think of any alternatives to this.

**Review by joeauyeung — DISMISSED** [review]
Tested and working. I have two unblocking comments.

**Review by keithwillcode — DISMISSED** [review]
Type checks are failing

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx (hunk 3 above)
Comment (sean-brydon) [reply packages/features/users/components/UserTable/DeleteMemberModal.tsx:29]: I'd say so - yeah. Helps other pages be invalidated that use this value from cache

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx (hunk 4 above)
Comment (sean-brydon) [reply packages/features/users/components/UserTable/DeleteMemberModal.tsx:23]: Only way to do this optimistically :/

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx [hunk 5]
```diff
@@ -13,15 +13,33 @@ export function DeleteMemberModal({ state, dispatch }: { state: State; dispatch:
...
+      dispatch({ type: "CLOSE_MODAL" });
     },
-    async onError(err) {
+    async onError(err, _, context) {
+      if (context?.previousValue) {
+        utils.viewer.organizations.listMembers.setInfiniteData(
+          { limit: 10, searchTerm: "" },
+          context.previousValue
+        );
+      }
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/DeleteMemberModal.tsx:42]:
```suggestion
```
Remove onError previousValue as we only do the above optimisitic update on succcess

File: packages/features/users/components/UserTable/DeleteMemberModal.tsx [hunk 6]
```diff
@@ -13,15 +13,26 @@ export function DeleteMemberModal({ state, dispatch }: { state: State; dispatch:
...
+      });
+
+      // Existing invalidations
+      Promise.all([utils.viewer.teams.get.invalidate(), utils.viewer.eventTypes.invalidate()]);
+
       showToast(t("success"), "success");
+
+      // Close the modal after successful deletion
+      dispatch({ type: "CLOSE_MODAL" });
     },
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/DeleteMemberModal.tsx:35]:
```suggestion
    }
```

**Review by emrysal — APPROVED** [review]
@joeauyeung Excellent review; Happy with the responses to your questions also so 👍

**Pull request merged**